### ISSUE TITLE
Streaming Inference Performance Gap vs Validation with Multi-Lookahead Training

### ISSUE THREAD

**Issue by HemendraSharma04** [NVIDIA-NeMo/NeMo#14930]
## Streaming Inference Performance Gap vs Validation with Multi-Lookahead Training

**Model:** FastConformer-Hybrid-RNNT-CTC (Cache-Aware Streaming)  
**NeMo Version:** 2.4.0  
**Issue:** Streaming inference WER significantly worse than validation WER, even with trained lookahead sizes

---

**Describe the bug**

Cache-aware streaming FastConformer-Hybrid-RNNT-CTC model trained with multi-lookahead training shows significantly worse WER in streaming inference compared to validation, even when using lookahead sizes the model was explicitly trained for. Validation with 80ms lookahead achieves ~10-12% WER, but streaming inference with 480ms lookahead (whose att_context_probs is set at 30% ) achieves ~21% WER. Streaming with 80ms lookahead completely fails and misses most words.

As written in training config the first element of `att_context_size` will be used in validation so 80ms would have been used , so we would expect similar results in inference with 80ms lookahead
```
For multi-lookahead models, you may specify a list of context sizes. During the training, different context sizes would be used randomly with the distribution specified by att_context_probs.
The first item in the list would be the default during test/validation/inference
```

**Steps/Code to reproduce bug**

1. **Train model with multi-lookahead configuration:**

- I referred this training config [training_file](https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/conf/fastconformer/hybrid_cache_aware_streaming/fastconformer_hybrid_transducer_ctc_bpe_streaming.yaml)

```yaml
encoder:
  att_context_size: [[70, 1], [70, 4], [70, 6]] # 80ms, 320ms, 480ms lookaheads
  att_context_style: chunked_limited
  att_context_probs: [0.4, 0.3, 0.3] # 40% on 80ms, 30% on 320ms, 30% on 480ms
  subsampling: dw_striding
  subsampling_factor: 8
  causal_downsampling: true
  conv_context_size: causal
```

2. **Run validation (achieves good performance):**

```python
import nemo.collections.asr as nemo_asr
from omegaconf import open_dict

model = nemo_asr.models.ASRModel.restore_from("model.nemo")
model.eval()

with open_dict(model.cfg):
    model.cfg.validation_ds.manifest_filepath = "val_manifest.json"
    model.cfg.validation_ds.batch_size = 1

model.setup_validation_data(model.cfg.validation_ds)

for batch_idx, batch in enumerate(model._validation_dl):
    logs = model.validation_pass(batch, batch_idx, 0)
    print(f"WER: {logs['val_wer']:.4f}")  # ~10-12%
```

3. **Run streaming inference (poor performance):**

- For inference i referred these two files :  [script](https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/asr_cache_aware_streaming/speech_to_text_cache_aware_streaming_infer.py) and [demo](https://github.com/NVIDIA-NeMo/NeMo/blob/main/tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb)

```python
import nemo.collections.asr as nemo_asr
import numpy as np
import torch
from omegaconf import OmegaConf
from nemo.collections.asr.models.hybrid_rnnt_ctc_bpe_models import EncDecHybridRNNTCTCBPEModel
import scipy.io.wavfile

model = nemo_asr.models.ASRModel.restore_from("model.nemo")

# Setup streaming with 480ms lookahead (trained for 30% of time)
left_context = model.encoder.att_context_size[0]
model.encoder.set_default_att_context_size([left_context, 6])  # 480ms
model.encoder.setup_streaming_params()
model.change_decoding_strategy(decoder_type="rnnt")
model.eval()

# Initialize preprocessor (use model's config to ensure exact match)
cfg = OmegaConf.create(model._cfg.preprocessor)
OmegaConf.set_struct(cfg, False)
cfg.dither = 0.00001  # Match training config exactly
cfg.pad_to = 0
cfg.normalize = "NA"  # Match training config exactly
preprocessor = EncDecHybridRNNTCTCBPEModel.from_config_dict(cfg)
preprocessor.to(model.device)

# Initialize caches
cache_last_channel, cache_last_time, cache_last_channel_len = \
    model.encoder.get_initial_cache_state(batch_size=1)
previous_hypotheses = None
pred_out_stream = None

# Process audio in chunks
sr, audio = scipy.io.wavfile.read("audio.wav")
chunk_size = int(0.16 * sr)

for i in range(0, len(audio), chunk_size):
    chunk = audio[i:i + chunk_size].astype(np.float32) / 32768.0
    audio_signal = torch.from_numpy(chunk).unsqueeze_(0).to(model.device)
    audio_signal_len = torch.Tensor([chunk.shape[0]]).to(model.device)

    processed_signal, processed_signal_length = preprocessor(
        input_signal=audio_signal, length=audio_signal_len
    )

    with torch.no_grad():
        (pred_out_stream, texts, cache_last_channel, cache_last_time,
         cache_last_channel_len, previous_hypotheses) = \
            model.conformer_stream_step(
                processed_signal=processed_signal,
                processed_signal_length=processed_signal_length,
                cache_last_channel=cache_last_channel,
                cache_last_time=cache_last_time,
                cache_last_channel_len=cache_last_channel_len,
                keep_all_outputs=False,
                previous_hypotheses=previous_hypotheses,
                previous_pred_out=pred_out_stream,
                return_transcription=True
            )
    # Result: ~21% WER (significantly worse than validation)
```

**Expected behavior**

Streaming inference with 80ms lookahead should match or closely approach validation performance (~10-12% WER), since the model was trained 40% of the time with this exact lookahead configuration. The performance gap between validation and streaming should be minimal for a cache-aware streaming model.

**Environment overview (please complete the following information)**

- GPU : L4 
- Method of NeMo install: uv
- NeMo version: 2.4.0

**Environment details**

- OS version: ubuntu 22.04
- PyTorch version: 2.x (compatible with NeMo 2.4.0)
- Python version: 3.10+

**Additional context**

- **Model**: FastConformer-Hybrid-Transducer-CTC-BPE-Streaming
- **Model size**: ~115M parameters
- **Training data**: ~6000 hours
- **Sample rate**: 8kHz
- **Encoder**: 17-layer Conformer with d_model=512
- **Losses**: 60% RNNT + 10% InterCTC + 30% CTC

**Performance observations:**

- ✅ Validation with 80ms: ~10-12% WER (excellent)
- ❌ Streaming with 80ms: Missing most words (complete failure)
- ⚠️ Streaming with 480ms: ~21% WER 

**Questions:**

1. Is this performance gap expected for cache-aware streaming models with multi-lookahead training?
2. Why does 80ms lookahead work in validation but fail completely in streaming?
3. What are the recommended streaming inference settings for this training configuration?
4. Should we retrain with a different approach or is this an inference configuration issue?

**Comment by KunalDhawan** [comment]
Hi @HemendraSharma04, thanks for sharing the details. The bad inference accuracy is certainly off and not expected behavior. Let's try to debug the issue. 

Could you answer these questions for me please:
1) Is the validation and inference data from the same distribution? Could you share more details about the data (size, source, quality etc)?
2) Can you try using the following script for inference - https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/asr_cache_aware_streaming/speech_to_text_cache_aware_streaming_infer.py

**Comment by nithinraok** [comment]
As there has been no response, I am closing this as completed. Please feel free to reopen it if the suggested solution does not work.
